// docxtemplater/compiledTemplate.js
var CompiledTemplate;

CompiledTemplate = CompiledTemplate = (function() {
  function CompiledTemplate() {
    this.compiled = [];
    this;
  }

  CompiledTemplate.prototype.prependText = function(text) {
    this.compiled.unshift(text);
    return this;
  };

  CompiledTemplate.prototype.appendTag = function(compiledTag) {
    if (!compiledTag) {
      throw new Error("compiled tag empty!");
    }
    this.compiled = this.compiled.concat(compiledTag.compiled);
    return this;
  };

  CompiledTemplate.prototype.appendRaw = function(tag) {
    this.compiled.push({
      type: 'raw',
      tag: tag
    });
    return this;
  };

  CompiledTemplate.prototype.appendText = function(text) {
    if (text !== '') {
      this.compiled.push(text);
    }
    return this;
  };

  CompiledTemplate.prototype.appendSubTemplate = function(subTemplate, tag, inverted) {
    if (!subTemplate) {
      throw new Error("subTemplate empty!");
    }
    return this.compiled.push({
      type: 'loop',
      tag: tag,
      inverted: inverted,
      template: subTemplate
    });
  };

  return CompiledTemplate;

})();

module.exports = CompiledTemplate;
